matrix-room.c

```diff
@@ -350,6 +350,135 @@ void _event_send_bad_response(MatrixConnectionData *ma, gpointer user_data,
+             sid->event->txn_id, sid->event->content, _event_send_complete,
+             _event_send_error, _event_send_bad_response, sid->conv);
+    purple_conversation_set_data(sid->conv, PURPLE_CONV_DATA_ACTIVE_SEND,
+                    fetch_data);
+    purple_imgstore_unref(image);
+}
+
+static void _image_upload_bad_response(MatrixConnectionData *ma, gpointer user_data,
+            int http_response_code, JsonNode *json_root)
+{
+    struct SendImageData *sid = user_data;
+    PurpleStoredImage *image = purple_imgstore_find_by_id(sid->imgstore_id);
+
+    matrix_api_bad_response(ma, sid->conv, http_response_code, json_root);
+    purple_imgstore_unref(image);
+    purple_conversation_set_data(sid->conv, PURPLE_CONV_DATA_ACTIVE_SEND, NULL);
+    /* More clear up with the message? */
+}
+
+void _image_upload_error(MatrixConnectionData *ma, gpointer user_data,
+            const gchar *error_message)
+{
+    struct SendImageData *sid = user_data;
+    PurpleStoredImage *image = purple_imgstore_find_by_id(sid->imgstore_id);
+
+    matrix_api_error(ma, sid->conv, error_message);
+    purple_imgstore_unref(image);
+    purple_conversation_set_data(sid->conv, PURPLE_CONV_DATA_ACTIVE_SEND, NULL);
+    /* More clear up with the message? */
+}
+
+/**
+ * Return a mimetype based on some info; this should get replaced
+ * with a glib/gio/gcontent_type_guess call if we can include it,
+ * all other plugins do this manually.
+ */
+static const char *type_guess(PurpleStoredImage *image)
+{
+    /* Copied off the code in libpurple's jabber module */
+    const char *ext = purple_imgstore_get_extension(image);
+
+    if (strcmp(ext, "png") == 0) {
+      return "image/png";
+    } else if (strcmp(ext, "gif") == 0) {
+      return "image/gif";
+    } else if (strcmp(ext, "jpg") == 0) {
+      return "image/jpeg";
+    } else if (strcmp(ext, "tif") == 0) {
+      return "image/tif";
+    } else {
+      return "image/x-icon"; /* or something... */
+    }
+}
+
+/**
+ * Called back by _send_queued_event for an image.
+ */
+static void _send_image_hook(void *opaque, MatrixRoomEvent *event)
+{
+    MatrixApiRequestData *fetch_data;
+    PurpleConversation *conv = opaque;
+    PurpleConnection *pc = conv->account->gc;
+    MatrixConnectionData *acct = purple_connection_get_protocol_data(pc);
+    struct SendImageData *sid = event->hook_data;
+    int imgstore_id = sid->imgstore_id;
+    PurpleStoredImage *image = purple_imgstore_find_by_id(imgstore_id);
+    size_t imgsize;
+    const char *filename;
+    const char *ctype;
+    gconstpointer imgdata;
+
+    if (!image)
+        return;
+
+    imgsize = purple_imgstore_get_size(image);
+    filename = purple_imgstore_get_filename(image);
+    imgdata = purple_imgstore_get_data(image);
+    ctype = type_guess(image);
+
+    purple_debug_info("matrixprpl", "%s: image id %d for %s (type: %s)\n",
+            __func__,
+            sid->imgstore_id, filename, ctype);
+
+    sid->event = event;
```

Comment:
ugh. It feels like you have ended up jumping through some hoops so that you can share the one SendImageData structure between `event->hook_data` and the `matrix_api_upload_file` data. I think you should allocate a new structure here and pass that into `matrix_api_upload_file`.